File: bindings/python/src/graphqlite/graph/nodes.py
"""Node operations mixin for Graph class."""

import json
from typing import Any, Optional

from ._base import BaseMixin


class NodesMixin(BaseMixin):
    """Mixin providing node CRUD operations."""

    def has_node(self, node_id: str) -> bool:
        """
        Check if a node exists.

        Args:
            node_id: The node's id property value

        Returns:
            True if node exists, False otherwise
        """
        result = self._conn.cypher(
            "MATCH (n {id: $id}) RETURN count(n) AS cnt",
            params={"id": node_id},
        )
        if len(result) == 0:
            return False
        cnt = result[0].get("cnt", 0)
        return int(cnt) > 0 if cnt else False

    def get_node(self, node_id: str) -> Optional[dict]:
        """
        Get a node by ID.

        Args:
            node_id: The node's id property value

        Returns:
            Node dict with 'id', 'labels', 'properties' or None if not found
        """
        result = self._conn.cypher(
            "MATCH (n {id: $id}) RETURN n",
            params={"id": node_id},
        )
        if len(result) == 0:
            return None
        return result[0].get("n")

    def upsert_node(
        self,
        node_id: str,
        node_data: dict[str, Any],
        label: str = "Entity"
    ) -> None:
        """
        Create or update a node.

        If a node with the given id exists, its properties are updated.
        Otherwise, a new node is created.

        Args:
            node_id: Unique identifier for the node (stored as 'id' property)
            node_data: Dictionary of properties to set
            label: Node label (only used on creation)
        """
        props = {"id": node_id, **node_data}

        if self.has_node(node_id):
            # Update existing node
            for k, v in node_data.items():
                self._conn.cypher(
                    f"MATCH (n {{id: $id}}) "
                    f"SET n.{k} = $val RETURN n",
                    params={"id": node_id, "val": v},
                )
        else:
            # Create new node
            prop_str = self._format_props(props)
            self._conn.cypher(f"CREATE (n:{label} {{{prop_str}}})")

    def delete_node(self, node_id: str) -> None:
        """
        Delete a node and its relationships.

        Args:
            node_id: The node's id property value
        """
        self._conn.cypher(
            "MATCH (n {id: $id}) DETACH DELETE n",
            params={"id": node_id},
        )

    def get_all_nodes(self, label: Optional[str] = None) -> list[dict]:
        """
        Get all nodes, optionally filtered by label.

        Args:
            label: Optional label to filter by

        Returns:
            List of node dicts
        """
        if label:
            result = self._conn.cypher(f"MATCH (n:{label}) RETURN n")
        else:
            result = self._conn.cypher("MATCH (n) RETURN n")

        nodes = []
        for row in result:
            if "result" in row and isinstance(row["result"], str):
                try:
                    parsed = json.loads(row["result"])
                    for item in parsed:
                        if "n" in item:
                            nodes.append(item["n"])
                except json.JSONDecodeError:
                    pass
            elif "n" in row and row["n"]:
                nodes.append(row["n"])
        return nodes

File: bindings/rust/src/graph/nodes.rs
//! Node operations for Graph.

use super::Graph;
use crate::utils::{escape_string, PropertyValue};
use crate::{Result, Value};

impl Graph {
    /// Check if a node with the given ID exists.
    pub fn has_node(&self, node_id: &str) -> Result<bool> {
        let result = self
            .connection()
            .cypher_builder("MATCH (n {id: $id}) RETURN count(n) AS cnt")
            .param("id", node_id)
            .run()?;
        if result.is_empty() {
            return Ok(false);
        }
        let cnt: i64 = result[0].get("cnt").unwrap_or(0);
        Ok(cnt > 0)
    }

    /// Get a node by ID.
    ///
    /// Returns the node as a [`Value`], or `None` if not found.
    pub fn get_node(&self, node_id: &str) -> Result<Option<Value>> {
        let result = self
            .connection()
            .cypher_builder("MATCH (n {id: $id}) RETURN n")
            .param("id", node_id)
            .run()?;
        if result.is_empty() {
            return Ok(None);
        }
        Ok(result[0].get_value("n").cloned())
    }

    /// Create or update a node.
    ///
    /// If a node with the given id exists, its properties are updated.
    /// Otherwise, a new node is created.
    pub fn upsert_node<I, K, V>(&self, node_id: &str, props: I, label: &str) -> Result<()>
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: Into<PropertyValue>,
    {
        let props: Vec<(String, PropertyValue)> = props
            .into_iter()
            .map(|(k, v)| (k.as_ref().to_string(), v.into()))
            .collect();

        if self.has_node(node_id)? {
            // Update existing node
            for (k, v) in props {
                let query = format!(
                    "MATCH (n {{id: $id}}) SET n.{} = {} RETURN n",
                    k,
                    v.to_cypher()
                );
                self.connection()
                    .cypher_builder(&query)
                    .param("id", node_id)
                    .run()?;
            }
        } else {
            // Create new node
            let mut prop_parts = vec![format!("id: '{}'", escape_string(node_id))];
            for (k, v) in props {
                prop_parts.push(format!("{}: {}", k, v.to_cypher()));
            }
            let prop_str = prop_parts.join(", ");
            let query = format!("CREATE (n:{} {{{}}})", label, prop_str);
            self.connection().cypher(&query)?;
        }
        Ok(())
    }

    /// Delete a node and all its relationships.
    pub fn delete_node(&self, node_id: &str) -> Result<()> {
        self.connection()
            .cypher_builder("MATCH (n {id: $id}) DETACH DELETE n")
            .param("id", node_id)
            .run()?;
        Ok(())
    }

    /// Get all nodes, optionally filtered by label.
    pub fn get_all_nodes(&self, label: Option<&str>) -> Result<Vec<Value>> {
        let query = match label {
            Some(l) => format!("MATCH (n:{}) RETURN n", l),
            None => "MATCH (n) RETURN n".to_string(),
        };
        let result = self.connection().cypher(&query)?;
        let mut nodes = Vec::new();
        for row in result.iter() {
            if let Some(n) = row.get_value("n") {
                nodes.push(n.clone());
            }
        }
        Ok(nodes)
    }
}
